crates/ruff_db/src/vendored/path.rs

```diff
@@ -1,111 +1,91 @@
-use std::ops::Deref;
+use std::fmt;
 use std::path;
 
-use camino::{Utf8Components, Utf8Path, Utf8PathBuf};
+#[derive(Debug)]
+pub struct UnsupportedComponentError(String);
 
-#[repr(transparent)]
-#[derive(Debug, Eq, PartialEq, Hash)]
-pub struct VendoredPath(Utf8Path);
-
-impl VendoredPath {
-    pub fn new(path: &(impl AsRef<Utf8Path> + ?Sized)) -> &Self {
-        let path = path.as_ref();
-        // SAFETY: VendoredPath is marked as #[repr(transparent)] so the conversion from a
-        // *const Utf8Path to a *const VendoredPath is valid.
-        unsafe { &*(path as *const Utf8Path as *const VendoredPath) }
-    }
-
-    pub fn to_path_buf(&self) -> VendoredPathBuf {
-        VendoredPathBuf(self.0.to_path_buf())
-    }
-
-    pub fn as_str(&self) -> &str {
-        self.0.as_str()
-    }
-
-    pub fn as_std_path(&self) -> &path::Path {
-        self.0.as_std_path()
-    }
-
-    pub fn components(&self) -> Utf8Components {
-        self.0.components()
+impl fmt::Display for UnsupportedComponentError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Unsupported component in a vendored path: {:?}", self.0)
     }
 }
 
-#[repr(transparent)]
-#[derive(Debug, Eq, PartialEq, Clone, Hash)]
-pub struct VendoredPathBuf(Utf8PathBuf);
+impl std::error::Error for UnsupportedComponentError {}
 
-impl Default for VendoredPathBuf {
-    fn default() -> Self {
-        Self::new()
-    }
-}
+#[derive(Debug, Eq, PartialEq, Clone, Hash, Default)]
+pub struct VendoredPathBuf(String);
 
 impl VendoredPathBuf {
-    pub fn new() -> Self {
-        Self(Utf8PathBuf::new())
-    }
+    pub fn new(path: &camino::Utf8Path) -> Result<Self, UnsupportedComponentError> {
+        // Allow the `RootDir` component, but only if it is at the very start of the string.
+        let mut components = path.components().peekable();
+        if let Some(camino::Utf8Component::RootDir) = components.peek() {
+            components.next();
+        }
 
-    pub fn as_path(&self) -> &VendoredPath {
-        VendoredPath::new(&self.0)
-    }
-}
+        let mut normalized_parts = Vec::new();
 
-impl AsRef<VendoredPath> for VendoredPathBuf {
-    fn as_ref(&self) -> &VendoredPath {
-        self.as_path()
+        for component in components {
+            match component {
+                camino::Utf8Component::Normal(part) => normalized_parts.push(part),
+                camino::Utf8Component::CurDir => continue,
+                camino::Utf8Component::ParentDir => {
+                    normalized_parts.pop();
```

Comment:
This will yield an incorrect result if you have ../ and normalized_parts is empty

Reply:
Good point. Fixed in https://github.com/astral-sh/ruff/pull/11989/commits/0775bc715c1b905c265e496ea74e765e11bccc89.